pyteal/stack_frame.py

```diff
@@ -318,46 +331,96 @@ def _best_frame_as_pyteal_frame(self) -> "PyTealFrame | None":
+        return f"NatalStackFrame({self._frames=})"
 
     def nodes(self) -> list[AST | None]:
         return [f.node for f in self._frames]
 
     @classmethod
-    def _walk_asts(cls, func: Callable[["Expr"], None], *exprs: "Expr") -> None:  # type: ignore
+    def _walk_asts(
+        cls,
+        func: Callable[["Expr"], bool],  # type: ignore[name-defined]
+        *exprs: "Expr",  # type: ignore[name-defined]
+        force_root_apply: bool = True,
+        exit_on_user_defined: bool = False,
+        exit_when_stop_signal: bool = False,
+    ) -> None:
         from pyteal.ast import (
             Assert,
             BinaryExpr,
             Cond,
             Expr,
+            Int,
+            MethodSignature,
             Return,
+            ScratchStackStore,
+            ScratchStore,
             Seq,
+            SubroutineCall,
             SubroutineDeclaration,
+            TxnaExpr,
+            UnaryExpr,
         )
+
         from pyteal.ast.frame import Proto
+        from pyteal.ast.return_ import ExitProgram
 
-        for expr in exprs:
-            e = cast(Expr, expr)
-            func(e)
+        for e in exprs:
+            if e.stack_frames.user_defined() and exit_on_user_defined:
+                continue
 
-            match e:
+            supported_type = True
+            expr = cast(Expr, e)
+            walker_args: list[Expr] = []
+            match expr:
                 case Assert():
-                    cls._walk_asts(func, *e.cond)
+                    walker_args = expr.cond
                 case BinaryExpr():
-                    cls._walk_asts(func, e.argLeft, e.argRight)
+                    walker_args = [expr.argLeft, expr.argRight]
                 case Cond():
-                    cls._walk_asts(func, *(y for x in e.args for y in x))
+                    walker_args = [y for x in expr.args for y in x]
+                case ExitProgram():
+                    walker_args = [expr.success]
                 case Proto():
-                    cls._walk_asts(func, e.mem_layout)
+                    walker_args = [expr.mem_layout]
                 case Seq():
-                    cls._walk_asts(func, *e.args)
+                    walker_args = cast(list[Expr], expr.args)
                 case Return():
-                    cls._walk_asts(func, e.value)
+                    if expr.value:
+                        walker_args = [expr.value]
                 case SubroutineDeclaration():
-                    cls._walk_asts(func, e.body)
-                case _:
-                    # TODO: implement more cases, but no need to error as this isn't used for functionality's sake.
+                    walker_args = [expr.body]
+                case ScratchStore():
+                    walker_args = [expr.value]
+                    if expr.index_expression:
+                        walker_args.append(expr.index_expression)
+                case SubroutineCall():
```

Comment:
qq: can you handle re-entrance of subroutine call caused by recursion?

Reply:
As explained in a discussion above, we know we can avoid infinite loops as we exit when the `stack_frames` param is observed to be identical to `self.stack_frames`. There is also 2 unit test that ensure this:
* https://github.com/algorand/pyteal/blob/6a49be7b4c74abe6fa08493bdb8e5585dabf0bca/tests/unit/sourcemap_constructs311_test.py#L1901
* https://github.com/algorand/pyteal/blob/6a49be7b4c74abe6fa08493bdb8e5585dabf0bca/tests/unit/sourcemap_constructs311_test.py#L1955

Reply:
This thread should be resolved by the discussion thread below.